nexus/src/app/ip_pool.rs

```diff
@@ -56,24 +71,125 @@ impl super::Nexus {
         opctx: &OpContext,
         pool_params: &params::IpPoolCreate,
     ) -> CreateResult<db::model::IpPool> {
-        let silo_id = match pool_params.clone().silo {
-            Some(silo) => {
-                let (.., authz_silo) = self
-                    .silo_lookup(&opctx, silo)?
-                    .lookup_for(authz::Action::Read)
-                    .await?;
-                Some(authz_silo.id())
-            }
-            _ => None,
-        };
-        let pool = db::model::IpPool::new(
-            &pool_params.identity,
-            silo_id,
-            pool_params.is_default,
-        );
+        let pool = db::model::IpPool::new(&pool_params.identity);
         self.db_datastore.ip_pool_create(opctx, pool).await
     }
 
+    // TODO: this is used by a developer user to see what IP pools they can use
+    // in their silo, so it would be nice to say which one is the default
+
+    /// List IP pools in current silo
+    pub(crate) async fn silo_ip_pools_list(
+        &self,
+        opctx: &OpContext,
+        pagparams: &PaginatedBy<'_>,
+    ) -> ListResultVec<db::model::IpPool> {
+        self.db_datastore.silo_ip_pools_list(opctx, pagparams).await
+    }
+
+    // Look up pool by name or ID, but only return it if it's linked to the
+    // current silo
+    pub async fn silo_ip_pool_fetch<'a>(
+        &'a self,
+        opctx: &'a OpContext,
+        pool: &'a NameOrId,
+    ) -> LookupResult<db::model::IpPool> {
+        let (authz_pool, pool) =
+            self.ip_pool_lookup(opctx, pool)?.fetch().await?;
+
+        // 404 if no link is found in the current silo
+        let link = self.db_datastore.ip_pool_fetch_link(opctx, pool.id()).await;
+        if link.is_err() {
+            return Err(authz_pool.not_found());
+        }
+
+        Ok(pool)
+    }
+
+    pub(crate) async fn ip_pool_association_list(
+        &self,
+        opctx: &OpContext,
+        pool_lookup: &lookup::IpPool<'_>,
+        pagparams: &DataPageParams<'_, Uuid>,
+    ) -> ListResultVec<db::model::IpPoolResource> {
+        // TODO: is this the right action to check?
+        let (.., authz_pool) =
+            pool_lookup.lookup_for(authz::Action::ListChildren).await?;
+        self.db_datastore
+            .ip_pool_association_list(opctx, &authz_pool, pagparams)
+            .await
+    }
+
+    pub(crate) async fn ip_pool_associate_resource(
+        &self,
+        opctx: &OpContext,
+        pool_lookup: &lookup::IpPool<'_>,
+        silo_link: &params::IpPoolSiloLink,
+    ) -> CreateResult<db::model::IpPoolResource> {
+        let (.., authz_pool) =
+            pool_lookup.lookup_for(authz::Action::Modify).await?;
+        let (silo,) = self
+            .silo_lookup(&opctx, silo_link.silo.clone())?
+            .lookup_for(authz::Action::Read)
+            .await?;
```

Comment:
Well this can't be right. Plus it's different from the perm check on dissociate.

Reply:
I think for now it certainly seems fine. We could go through and create a synthetic resource for the link which is probably the correct means of handling this. Worth an issue to investigate that.